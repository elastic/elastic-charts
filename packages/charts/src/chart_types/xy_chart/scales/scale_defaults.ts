/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { ScaleType } from '../../../scales/constants';

/** @internal */
export const X_SCALE_DEFAULT = {
  type: ScaleType.Ordinal,
  nice: false,
  desiredTickCount: 10,
};

/** @internal */
export const Y_SCALE_DEFAULT = {
  type: ScaleType.Linear,
  nice: false,
  desiredTickCount: 5,
  constrainDomainPadding: undefined,
  domainPixelPadding: 0,
  logBase: undefined,
  logMinLimit: undefined,
};
