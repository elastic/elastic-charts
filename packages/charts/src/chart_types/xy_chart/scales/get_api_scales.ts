/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { X_SCALE_DEFAULT, Y_SCALE_DEFAULT } from './scale_defaults';
import { ScaleContinuousType } from '../../../scales';
import { BasicSeriesSpec, XScaleType } from '../utils/specs';

/** @internal */
export function getXScaleTypeFromSpec(type?: BasicSeriesSpec['xScaleType']): XScaleType {
  return type ?? X_SCALE_DEFAULT.type;
}

/** @internal */
export function getXNiceFromSpec(nice?: BasicSeriesSpec['xNice']): boolean {
  return nice ?? X_SCALE_DEFAULT.nice;
}

/** @internal */
export function getYScaleTypeFromSpec(type?: BasicSeriesSpec['yScaleType']): ScaleContinuousType {
  return type ?? Y_SCALE_DEFAULT.type;
}

/** @internal */
export function getYNiceFromSpec(nice?: BasicSeriesSpec['yNice']): boolean {
  return nice ?? Y_SCALE_DEFAULT.nice;
}
