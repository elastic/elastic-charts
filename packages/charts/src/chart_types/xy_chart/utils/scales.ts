/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Scale, ScaleBand, ScaleContinuous } from '../../../scales';
import { ScaleType } from '../../../scales/constants';
import { LogBase } from '../../../scales/scale_continuous';
import { ContinuousDomain, Range } from '../../../utils/domain';
import { GroupId } from '../../../utils/ids';
import { XDomain, YDomain } from '../domains/types';

function getBandScaleRange(
  isInverse: boolean,
  isSingleValueHistogram: boolean,
  minRange: number,
  maxRange: number,
  bandwidth: number,
): {
  start: number;
  end: number;
} {
  const rangeEndOffset = isSingleValueHistogram ? 0 : bandwidth;
  const start = isInverse ? minRange - rangeEndOffset : minRange;
  const end = isInverse ? maxRange : maxRange - rangeEndOffset;
  return { start, end };
}

interface XScaleOptions {
  xDomain: XDomain;
  totalBarsInCluster: number;
  range: Range;
  barsPadding?: number;
  enableHistogramMode?: boolean;
  integersOnly?: boolean;
  logBase?: LogBase;
  logMinLimit?: number;
}

/**
 * Compute the x scale used to align geometries to the x axis.
 * @internal
 */
export function computeXScale(options: XScaleOptions): Scale {
  const { xDomain, totalBarsInCluster, range, barsPadding, enableHistogramMode, integersOnly } = options;
  const { type, nice, minInterval, domain, isBandScale, timeZone, logBase, desiredTickCount } = xDomain;
  const rangeDiff = Math.abs(range[1] - range[0]);
  const isInverse = range[1] < range[0];
  if (type === ScaleType.Ordinal) {
    const dividend = totalBarsInCluster > 0 ? totalBarsInCluster : 1;
    const bandwidth = rangeDiff / (domain.length * dividend);
    return new ScaleBand(domain, range, bandwidth, barsPadding);
  }
  if (isBandScale) {
    const [domainMin, domainMax] = domain as ContinuousDomain;
    const isSingleValueHistogram = !!enableHistogramMode && domainMax - domainMin === 0;

    const adjustedDomainMax = isSingleValueHistogram ? domainMin + minInterval : domainMax;
    const adjustedDomain = [domainMin, adjustedDomainMax];

    const intervalCount = (adjustedDomain[1] - adjustedDomain[0]) / minInterval;
    const intervalCountOffset = isSingleValueHistogram ? 0 : 1;
    const bandwidth = rangeDiff / (intervalCount + intervalCountOffset);
    const { start, end } = getBandScaleRange(isInverse, isSingleValueHistogram, range[0], range[1], bandwidth);

    return new ScaleContinuous(
      {
        type,
        domain: adjustedDomain,
        range: [start, end],
        nice,
      },
      {
        bandwidth: totalBarsInCluster > 0 ? bandwidth / totalBarsInCluster : bandwidth,
        minInterval,
        timeZone,
        totalBarsInCluster,
        barsPadding,
        desiredTickCount,
        isSingleValueHistogram,
        logBase,
      },
    );
  }
  return new ScaleContinuous(
    { type, domain, range, nice },
    {
      bandwidth: 0,
      minInterval,
      timeZone,
      totalBarsInCluster,
      barsPadding,
      desiredTickCount,
      integersOnly,
      logBase,
    },
  );
}

interface YScaleOptions {
  yDomains: YDomain[];
  range: Range;
  integersOnly?: boolean;
}

/**
 * Compute the y scales, one per groupId for the y axis.
 * @internal
 */
export function computeYScales(options: YScaleOptions): Map<GroupId, Scale> {
  const { yDomains, range, integersOnly } = options;
  return yDomains.reduce(
    (
      yScales,
      {
        type,
        nice,
        desiredTickCount,
        domain,
        groupId,
        logBase,
        logMinLimit,
        domainPixelPadding,
        constrainDomainPadding,
      },
    ) => {
      const yScale = new ScaleContinuous(
        {
          type,
          domain,
          range,
          nice,
        },
        {
          desiredTickCount,
          integersOnly,
          logBase,
          logMinLimit,
          domainPixelPadding,
          constrainDomainPadding,
        },
      );
      yScales.set(groupId, yScale);
      return yScales;
    },
    new Map<GroupId, Scale>(),
  );
}
