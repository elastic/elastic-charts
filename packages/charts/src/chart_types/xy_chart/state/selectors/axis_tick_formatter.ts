/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_spec';
import { Rotation } from '../../../../utils/common';
import { SpecId } from '../../../../utils/ids';
import { defaultTickFormatter, isXDomain } from '../../utils/axis_utils';
import { groupBy } from '../../utils/group_data_series';
import { AxisSpec } from '../../utils/specs';
import { getScaleConfigsFromSpecsSelector } from './get_api_scale_configs';
import { getAxisSpecsSelector, getSeriesSpecsSelector } from './get_specs';

/** @internal */
export type AxisLabelFormatter<V = unknown> = (value: V) => string;

/** @internal */
export type AxisLabelFormatters = { x: Map<SpecId, AxisLabelFormatter>; y: Map<SpecId, AxisLabelFormatter> };

/** @internal */
export const getAxisTickLabelFormatter = createCustomCachedSelector(
  [getSeriesSpecsSelector, getAxisSpecsSelector, getSettingsSpecSelector, getScaleConfigsFromSpecsSelector],
  (seriesSpecs, axesSpecs, { rotation }, scaleConfigs): AxisLabelFormatters => {
    const seriesByGroupId = groupBy(seriesSpecs, ['groupId'], false);
    const axesByGroupId = groupBy(axesSpecs, ['groupId'], false);
    const groupIds = [...new Set([...Object.keys(seriesByGroupId), ...Object.keys(axesByGroupId)])];
    const { timeZone } = scaleConfigs.x;
    // we need to do that by groupId to find the first Y spec formatter fallback
    return groupIds.reduce<AxisLabelFormatters>(
      (acc, groupId) => {
        const ySpecDataFormatter = (seriesByGroupId[groupId] ?? []).find(({ tickFormat }) => tickFormat)?.tickFormat;
        const axes = groupAxesByCartesianCoords(axesByGroupId[groupId] ?? [], rotation);
        axes.x.forEach((spec) => {
          acc.x.set(spec.id, (v) => (spec?.labelFormat ?? spec?.tickFormat ?? defaultTickFormatter)(v, { timeZone }));
        });
        axes.y.forEach((spec) => {
          acc.y.set(spec.id, (v) =>
            (spec.labelFormat ?? spec.tickFormat ?? ySpecDataFormatter ?? defaultTickFormatter)(v, {}),
          );
        });
        return acc;
      },
      { x: new Map(), y: new Map() },
    );
  },
);

function groupAxesByCartesianCoords(sameGroupAxes: AxisSpec[], chartRotation: Rotation = 0) {
  return sameGroupAxes.reduce<{ x: AxisSpec[]; y: AxisSpec[] }>(
    (acc, spec) => {
      acc[isXDomain(spec.position, chartRotation) ? 'x' : 'y'].push(spec);
      return acc;
    },
    { x: [], y: [] },
  );
}
