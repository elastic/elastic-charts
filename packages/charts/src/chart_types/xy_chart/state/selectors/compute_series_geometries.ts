/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getChartThemeSelector } from '../../../../state/selectors/get_chart_theme';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_spec';
import { withTextMeasure } from '../../../../utils/bbox/canvas_text_bbox_calculator';
import { computeSeriesGeometries } from '../utils/utils';
import { getFallBackTickFormatter } from './compute_axis_ticks_dimensions';
import { computeSeriesDomainsSelector } from './compute_series_domains';
import { computeSmallMultipleScalesSelector } from './compute_small_multiple_scales';
import { getSeriesColorsSelector } from './get_series_color_map';
import { getSeriesSpecsSelector, getAxisSpecsSelector } from './get_specs';
import { isHistogramModeEnabledSelector } from './is_histogram_mode_enabled';

/** @internal */
export const computeSeriesGeometriesSelector = createCustomCachedSelector(
  [
    getSeriesSpecsSelector,
    computeSeriesDomainsSelector,
    getSeriesColorsSelector,
    getChartThemeSelector,
    getSettingsSpecSelector,
    getAxisSpecsSelector,
    computeSmallMultipleScalesSelector,
    isHistogramModeEnabledSelector,
    getFallBackTickFormatter,
  ],
  (specs, domain, colors, theme, settings, axis, smScales, isHistogram, fallbackFormatter) => {
    return withTextMeasure((measureText) =>
      computeSeriesGeometries(
        specs,
        domain,
        colors,
        theme,
        settings,
        axis,
        smScales,
        isHistogram,
        fallbackFormatter,
        measureText,
      ),
    );
  },
);
