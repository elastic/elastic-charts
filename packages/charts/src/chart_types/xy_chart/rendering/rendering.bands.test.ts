/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { MockGlobalSpec, MockSeriesSpec } from '../../../mocks/specs';
import { MockStore } from '../../../mocks/store';
import { ScaleType } from '../../../scales/constants';
import { computeSeriesGeometriesSelector } from '../state/selectors/compute_series_geometries';

const SPEC_ID = 'spec_1';
const GROUP_ID = 'group_1';

describe('Rendering bands - areas', () => {
  describe('Single band area chart', () => {
    const pointSeriesSpec = MockSeriesSpec.area({
      id: SPEC_ID,
      data: [
        [0, 2, 10],
        [1, 3, 5],
      ],
      xAccessor: 0,
      y0Accessors: [1],
      yAccessors: [2],
      xScaleType: ScaleType.Ordinal,
      yScaleType: ScaleType.Linear,
    });
    const store = MockStore.default();
    const settings = MockGlobalSpec.settingsNoMargins({ theme: { colors: { vizColors: ['red', 'blue'] } } });
    MockStore.addSpecs([pointSeriesSpec, settings], store);
    const {
      geometries: { areas },
    } = computeSeriesGeometriesSelector(store.getState());

    test('Can render upper and lower lines and area paths', () => {
      const [
        {
          value: { lines, area, color, seriesIdentifier, transform },
        },
      ] = areas;
      expect(lines.length).toBe(2);
      expect(lines[0]).toBe('M0,0L50,50');
      expect(lines[1]).toBe('M0,80L50,70');
      expect(area).toBe('M0,0L50,50L50,70L0,80Z');
      expect(color).toBe('red');
      expect(seriesIdentifier.seriesKeys).toEqual([2]);
      expect(seriesIdentifier.specId).toEqual(SPEC_ID);
      expect(transform).toEqual({ x: 25, y: 0 });
    });

    test('Can render two points', () => {
      const [
        {
          value: { points },
        },
      ] = areas;
      expect(points).toMatchSnapshot();
    });
  });
  describe('Single band area chart with null values', () => {
    const pointSeriesSpec = MockSeriesSpec.area({
      id: SPEC_ID,
      groupId: GROUP_ID,
      data: [
        [0, 2, 10],
        [1, 2, null],
        [2, 3, 5],
        [3, 3, 5],
      ],
      xAccessor: 0,
      y0Accessors: [1],
      yAccessors: [2],
      xScaleType: ScaleType.Ordinal,
      yScaleType: ScaleType.Linear,
    });
    const store = MockStore.default();
    const settings = MockGlobalSpec.settingsNoMargins({ theme: { colors: { vizColors: ['red', 'blue'] } } });
    MockStore.addSpecs([pointSeriesSpec, settings], store);
    const {
      geometries: { areas },
    } = computeSeriesGeometriesSelector(store.getState());

    test('Can render upper and lower lines and area paths', () => {
      const [
        {
          value: { lines, area, color, seriesIdentifier, transform },
        },
      ] = areas;
      expect(lines.length).toBe(2);
      expect(lines[0]).toBe('M0,0ZM50,50L75,50');
      expect(lines[1]).toBe('M0,80ZM50,70L75,70');
      expect(area).toBe('M0,0L0,80ZM50,50L75,50L75,70L50,70Z');
      expect(color).toBe('red');
      expect(seriesIdentifier.seriesKeys).toEqual([2]);
      expect(seriesIdentifier.specId).toEqual(SPEC_ID);
      expect(transform).toEqual({ x: 12.5, y: 0 });
    });

    test('Can render two points', () => {
      const [
        {
          value: { points },
        },
      ] = areas;
      expect(points).toMatchSnapshot();
    });
  });
  describe('Single series band bar chart - ordinal', () => {
    const barSeriesSpec = MockSeriesSpec.bar({
      id: SPEC_ID,
      groupId: GROUP_ID,
      data: [
        [0, 2, 10],
        [1, 3, null],
        [2, 3, 5],
        [3, 4, 8],
      ],
      xAccessor: 0,
      y0Accessors: [1],
      yAccessors: [2],
      xScaleType: ScaleType.Ordinal,
      yScaleType: ScaleType.Linear,
    });
    const store = MockStore.default();
    const settings = MockGlobalSpec.settingsNoMargins({ theme: { colors: { vizColors: ['red', 'blue'] } } });
    MockStore.addSpecs([barSeriesSpec, settings], store);
    const {
      geometries: {
        bars: [{ value: bars }],
      },
    } = computeSeriesGeometriesSelector(store.getState());

    test('Can render two bars', () => {
      expect(bars).toMatchSnapshot();
    });
  });
});
