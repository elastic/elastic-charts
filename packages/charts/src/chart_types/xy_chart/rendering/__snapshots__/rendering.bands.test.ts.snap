// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering bands - areas Single band area chart Can render two points 1`] = `
Array [
  Object {
    "color": "red",
    "orphan": false,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{__global__}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 25,
      "y": 0,
    },
    "value": Object {
      "accessor": "y0",
      "datum": Array [
        0,
        2,
        10,
      ],
      "mark": null,
      "x": 0,
      "y": 2,
    },
    "x": 0,
    "y": 80,
  },
  Object {
    "color": "red",
    "orphan": false,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{__global__}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 25,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        0,
        2,
        10,
      ],
      "mark": null,
      "x": 0,
      "y": 10,
    },
    "x": 0,
    "y": 0,
  },
  Object {
    "color": "red",
    "orphan": false,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{__global__}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 25,
      "y": 0,
    },
    "value": Object {
      "accessor": "y0",
      "datum": Array [
        1,
        3,
        5,
      ],
      "mark": null,
      "x": 1,
      "y": 3,
    },
    "x": 50,
    "y": 70,
  },
  Object {
    "color": "red",
    "orphan": false,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{__global__}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 25,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        1,
        3,
        5,
      ],
      "mark": null,
      "x": 1,
      "y": 5,
    },
    "x": 50,
    "y": 50,
  },
]
`;

exports[`Rendering bands - areas Single band area chart with null values Can render two points 1`] = `
Array [
  Object {
    "color": "red",
    "orphan": true,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 12.5,
      "y": 0,
    },
    "value": Object {
      "accessor": "y0",
      "datum": Array [
        0,
        2,
        10,
      ],
      "mark": null,
      "x": 0,
      "y": 2,
    },
    "x": 0,
    "y": 80,
  },
  Object {
    "color": "red",
    "orphan": true,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 12.5,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        0,
        2,
        10,
      ],
      "mark": null,
      "x": 0,
      "y": 10,
    },
    "x": 0,
    "y": 0,
  },
  Object {
    "color": "red",
    "orphan": false,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 12.5,
      "y": 0,
    },
    "value": Object {
      "accessor": "y0",
      "datum": Array [
        2,
        3,
        5,
      ],
      "mark": null,
      "x": 2,
      "y": 3,
    },
    "x": 50,
    "y": 70,
  },
  Object {
    "color": "red",
    "orphan": false,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 12.5,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        2,
        3,
        5,
      ],
      "mark": null,
      "x": 2,
      "y": 5,
    },
    "x": 50,
    "y": 50,
  },
  Object {
    "color": "red",
    "orphan": false,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 12.5,
      "y": 0,
    },
    "value": Object {
      "accessor": "y0",
      "datum": Array [
        3,
        3,
        5,
      ],
      "mark": null,
      "x": 3,
      "y": 3,
    },
    "x": 75,
    "y": 70,
  },
  Object {
    "color": "red",
    "orphan": false,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "radius": 2,
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "style": Object {
      "fill": Object {
        "color": Object {
          "b": 255,
          "g": 255,
          "opacity": 1,
          "r": 255,
        },
      },
      "shape": "circle",
      "stroke": Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "opacity": 1,
          "r": 255,
        },
        "width": 1,
      },
    },
    "transform": Object {
      "x": 12.5,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        3,
        3,
        5,
      ],
      "mark": null,
      "x": 3,
      "y": 5,
    },
    "x": 75,
    "y": 50,
  },
]
`;

exports[`Rendering bands - areas Single series band bar chart - ordinal Can render two bars 1`] = `
Array [
  Object {
    "color": "red",
    "displayValue": undefined,
    "height": 80,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        0,
        2,
        10,
      ],
      "mark": null,
      "x": 0,
      "y": 10,
    },
    "width": 25,
    "x": 0,
    "y": 0,
  },
  Object {
    "color": "red",
    "displayValue": undefined,
    "height": 20,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        2,
        3,
        5,
      ],
      "mark": null,
      "x": 2,
      "y": 5,
    },
    "width": 25,
    "x": 50,
    "y": 50,
  },
  Object {
    "color": "red",
    "displayValue": undefined,
    "height": 40,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{2}splitAccessors{}",
      "seriesKeys": Array [
        2,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 2,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        3,
        4,
        8,
      ],
      "mark": null,
      "x": 3,
      "y": 8,
    },
    "width": 25,
    "x": 75,
    "y": 20,
  },
]
`;
